compiler/rustc_typeck/src/check/generator_interior.rs

```diff
@@ -558,6 +560,12 @@ pub fn check_must_not_suspend_ty<'tcx>(
                 },
             )
         }
+        // If drop tracking is enabled, we want to look through references, since the referrent
+        // may not be considered live across the await point.
```

Comment:
Why don't we want to look through references when we are not in `drop_tracking` mode?

Reply:
the PR description says
> Unfortunately, this leads to duplicate warnings in some cases (e.g. [dedup.rs](https://cs.github.com/rust-lang/rust/blob/9a74608543d499bcc7dd505e195e8bfab9447315/src/test/ui/lint/must_not_suspend/dedup.rs#L4)), so we only use the new behavior when drop tracking is enabled.

Reply:
I did not see that this whole function could only emit a lint.
Anyway, why does looking through references cause duplicate warnings?

Reply:
Sorry for the delay, I finally have time to look at this again.

Without drop tracking, we use a scope-based analysis to decide what should be considered live across an await point, while the drop tracking version is more precise. With the scope based analysis, I think what's happening is we are seeing both a `&mut Umm` live across the await point and an `Umm` live across the await point. The first one doesn't trigger the lint because we currently don't see through the reference, but we do warn on the second.

With drop tracking, we don't see the `Umm` but instead only see `&mut Umm`. In order to warn, we need to look through the reference. The reason we don't see the `Umm` is because drop tracking handles temporary values differently, which leads to use not considering the field in `self`, but only the reference to it.